File: examples/go/apps/miniclaw/memory.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// MemoryActor — scoped KV + TupleSpace memory store.
// AuditEventActor — fire-and-forget audit trail with two-cursor polling (NanoClaw Pattern 3).
// AgentStateFSM — agent lifecycle state machine (GenFSM).
package main

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"

	"github.com/plexobject/plexspaces/sdks/go/plexspaces"
)

// ========================================================================
// MemoryActor
// ========================================================================

// MemoryActor provides scoped memory storage backed by KV (persistent) and
// TupleSpace (queryable). Supports global, agent, and session scope namespacing.
type MemoryActor struct {
	plexspaces.BaseActor
	MemoryCount int `json:"memory_count"`
}

func NewMemoryActor() plexspaces.Actor {
	a := &MemoryActor{}
	a.SetSelf(a)
	return a
}

func newMemoryActor() *MemoryActor {
	a := &MemoryActor{}
	a.SetSelf(a)
	return a
}

func (m *MemoryActor) Init(configJSON string) string {
	var config struct {
		ActorID string `json:"actor_id"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	m.SetRuntimeMetadata(config.ActorID)
	if err := host.PG().Join("svc:memory"); err != nil {
		host.Warn(fmt.Sprintf("MemoryActor: failed to join svc:memory: %v", err))
	}
	host.Info(fmt.Sprintf("MemoryActor Init actor_id=%s", config.ActorID))
	return ""
}

func (m *MemoryActor) Handle(fromActor, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	op := stringVal(p, "op", msgType)
	switch op {
	case "store_memory":
		return m.storeMemory(p)
	case "recall_memory":
		return m.recallMemory(p)
	case "list_memories":
		return m.listMemories(p)
	case "delete_memory":
		return m.deleteMemory(p)
	case "get_stats":
		return marshal(map[string]any{"status": "ok", "memory_count": m.MemoryCount})
	default:
		return marshal(map[string]any{"error": "unknown_op", "op": op})
	}
}

func (m *MemoryActor) storeMemory(p map[string]any) string {
	scope := stringVal(p, "scope", "global")
	scopeID := stringVal(p, "scope_id", "")
	key := stringVal(p, "key", "")
	value := stringVal(p, "value", "")
	if key == "" {
		return marshal(map[string]any{"error": "key is required"})
	}

	kvKey := fmt.Sprintf("mem:%s:%s:%s", scope, scopeID, key)
	host.KVPut(kvKey, value)
	tsOK := true
	if result := host.TS().Write([]any{"memory", scope, scopeID, key, value}); strings.HasPrefix(result, "ERROR:") {
		host.Warn(fmt.Sprintf("MemoryActor: TS write failed for key=%s: %s", key, result))
		tsOK = false
	}
	_ = tsOK
	m.MemoryCount++
	m.IncrCounter(host, "memories_stored")
	host.Info(fmt.Sprintf("MemoryActor: stored scope=%s key=%s", scope, key))
	return marshal(map[string]any{"status": "ok", "key": key, "scope": scope})
}

func (m *MemoryActor) recallMemory(p map[string]any) string {
	scope := stringVal(p, "scope", "global")
	scopeID := stringVal(p, "scope_id", "")
	query := stringVal(p, "query", "")

	tuples := host.TS().ReadAll([]any{"memory", scope, scopeID, nil, nil})
	memories := make([]any, 0)
	for _, t := range tuples {
		if len(t) < 5 {
			continue
		}
		memKey, _ := t[3].(string)
		memVal, _ := t[4].(string)
		if query == "" ||
			strings.Contains(strings.ToLower(memKey), strings.ToLower(query)) ||
			strings.Contains(strings.ToLower(memVal), strings.ToLower(query)) {
			memories = append(memories, map[string]any{
				"key":      memKey,
				"value":    memVal,
				"scope":    scope,
				"scope_id": scopeID,
			})
		}
	}
	return marshal(map[string]any{"status": "ok", "memories": memories, "count": len(memories), "query": query})
}

func (m *MemoryActor) listMemories(p map[string]any) string {
	scope := stringVal(p, "scope", "global")
	scopeID := stringVal(p, "scope_id", "")
	tuples := host.TS().ReadAll([]any{"memory", scope, scopeID, nil, nil})
	memories := make([]any, 0, len(tuples))
	for _, t := range tuples {
		if len(t) < 5 {
			continue
		}
		memKey, _ := t[3].(string)
		memVal, _ := t[4].(string)
		memories = append(memories, map[string]any{"key": memKey, "value": memVal, "scope": scope})
	}
	return marshal(map[string]any{"status": "ok", "memories": memories, "count": len(memories)})
}

func (m *MemoryActor) deleteMemory(p map[string]any) string {
	scope := stringVal(p, "scope", "global")
	scopeID := stringVal(p, "scope_id", "")
	key := stringVal(p, "key", "")
	if key == "" {
		return marshal(map[string]any{"error": "key is required"})
	}
	host.KVDelete(fmt.Sprintf("mem:%s:%s:%s", scope, scopeID, key))
	if m.MemoryCount > 0 {
		m.MemoryCount--
	}
	return marshal(map[string]any{"status": "ok", "key": key})
}

// ========================================================================
// AuditEventActor
// ========================================================================

// AuditEventActor records fire-and-forget audit events to TupleSpace and KV.
// Uses a monotonic Watermark (NanoClaw Pattern 3: Two-Cursor Message Processing)
// so consumers can poll with a per-consumer cursor for ordered, no-miss delivery.
// Declared with behavior_kind = "GenEvent" in app-config.toml.
type AuditEventActor struct {
	plexspaces.BaseActor
	EventsLogged  int    `json:"events_logged"`
	LastEventType string `json:"last_event_type"`
	Watermark     int64  `json:"watermark"`
}

func NewAuditEventActor() plexspaces.Actor {
	a := &AuditEventActor{}
	a.SetSelf(a)
	return a
}

func newAuditEventActor() *AuditEventActor {
	a := &AuditEventActor{}
	a.SetSelf(a)
	return a
}

func (a *AuditEventActor) Init(configJSON string) string {
	var config struct {
		ActorID string `json:"actor_id"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	a.SetRuntimeMetadata(config.ActorID)
	if err := host.PG().Join("svc:audit"); err != nil {
		host.Warn(fmt.Sprintf("AuditEventActor: failed to join svc:audit: %v", err))
	}
	host.Info(fmt.Sprintf("AuditEventActor Init actor_id=%s", config.ActorID))
	return ""
}

func (a *AuditEventActor) Handle(fromActor, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	op := stringVal(p, "op", msgType)
	switch op {
	case "log_event":
		return a.logEvent(p)
	case "query_events":
		return a.queryEvents(p)
	case "poll_events":
		return a.pollEvents(p)
	case "get_stats":
		return marshal(map[string]any{
			"status":          "ok",
			"events_logged":   a.EventsLogged,
			"last_event_type": a.LastEventType,
			"watermark":       a.Watermark,
		})
	default:
		if et := stringVal(p, "event_type", ""); et != "" {
			return a.logEvent(p)
		}
		return marshal(map[string]any{"ok": true})
	}
}

func (a *AuditEventActor) logEvent(p map[string]any) string {
	eventType := stringVal(p, "event_type", "unknown")
	detail := stringVal(p, "detail", "")
	ts := p["timestamp"]
	if ts == nil {
		ts = host.NowMs()
	}

	a.Watermark++
	seq := a.Watermark

	entry := map[string]any{
		"seq":        seq,
		"event_type": eventType,
		"detail":     detail,
		"source":     stringVal(p, "source", ""),
		"timestamp":  ts,
	}
	entryJSON, err := json.Marshal(entry)
	if err != nil {
		host.Warn(fmt.Sprintf("AuditEventActor: failed to marshal audit entry for event=%s: %v", eventType, err))
		return marshal(map[string]any{"ok": false, "error": "marshal_failed"})
	}

	host.KVPut(fmt.Sprintf("audit:seq:%d", seq), string(entryJSON))
	if result := host.TS().Write([]any{"audit", eventType, ts, string(entryJSON)}); strings.HasPrefix(result, "ERROR:") {
		host.Warn(fmt.Sprintf("AuditEventActor: TS write failed for event=%s: %s", eventType, result))
	}

	a.EventsLogged++
	a.LastEventType = eventType
	a.IncrCounter(host, "audit_events")
	return marshal(map[string]any{"ok": true, "seq": seq})
}

// pollEvents implements the two-cursor pattern: consumer sends its last-seen
// cursor; returns all events from (cursor+1) to current Watermark and the new
// cursor position. Consumers advance their own cursor in KV.
func (a *AuditEventActor) pollEvents(p map[string]any) string {
	consumerID := stringVal(p, "consumer_id", "default")
	limit := intVal(p, "limit", 50)

	cursorKey := "audit:cursor:" + consumerID
	cursorRaw := host.KVGet(cursorKey)
	var cursor int64
	if cursorRaw != "" {
		if n, err := strconv.ParseInt(cursorRaw, 10, 64); err == nil {
			cursor = n
		}
	}

	events := make([]any, 0)
	newCursor := cursor
	for seq := cursor + 1; seq <= a.Watermark && len(events) < limit; seq++ {
		raw := host.KVGet(fmt.Sprintf("audit:seq:%d", seq))
		if raw == "" {
			continue
		}
		var entry map[string]any
		if err := json.Unmarshal([]byte(raw), &entry); err == nil {
			events = append(events, entry)
			newCursor = seq
		}
	}

	host.KVPut(cursorKey, strconv.FormatInt(newCursor, 10))
	return marshal(map[string]any{
		"status":      "ok",
		"consumer_id": consumerID,
		"events":      events,
		"count":       len(events),
		"cursor":      newCursor,
		"watermark":   a.Watermark,
	})
}

func (a *AuditEventActor) queryEvents(p map[string]any) string {
	eventType := stringVal(p, "event_type", "")
	limit := intVal(p, "limit", 10)

	var tuples [][]any
	if eventType != "" {
		tuples = host.TS().ReadAll([]any{"audit", eventType, nil, nil})
	} else {
		tuples = host.TS().ReadAll([]any{"audit", nil, nil, nil})
	}
	if len(tuples) > limit {
		tuples = tuples[len(tuples)-limit:]
	}

	events := make([]any, 0, len(tuples))
	for _, t := range tuples {
		if len(t) < 4 {
			continue
		}
		et, _ := t[1].(string)
		ts := t[2]
		detail, _ := t[3].(string)
		events = append(events, map[string]any{"event_type": et, "timestamp": ts, "detail": detail})
	}
	return marshal(map[string]any{"status": "ok", "events": events, "count": len(events)})
}

// ========================================================================
// AgentStateFSM
// ========================================================================

// AgentStateFSM tracks the agent processing lifecycle.
// States: idle → processing → tool_executing → processing → responding → idle
// Any state can transition to error; error can recover to idle.
type AgentStateFSM struct {
	plexspaces.BaseActor
	FSMState        string `json:"fsm_state"`
	TransitionCount int    `json:"transition_count"`
}

func NewAgentStateFSM() plexspaces.Actor {
	a := &AgentStateFSM{FSMState: "idle"}
	a.SetSelf(a)
	return a
}

func newAgentStateFSM() *AgentStateFSM {
	a := &AgentStateFSM{FSMState: "idle"}
	a.SetSelf(a)
	return a
}

func (f *AgentStateFSM) Init(configJSON string) string {
	var config struct {
		ActorID string `json:"actor_id"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	f.SetRuntimeMetadata(config.ActorID)
	f.FSMState = "idle"
	if err := host.PG().Join("svc:agent_fsm"); err != nil {
		host.Warn(fmt.Sprintf("AgentStateFSM: failed to join svc:agent_fsm: %v", err))
	}
	host.Info(fmt.Sprintf("AgentStateFSM Init actor_id=%s", config.ActorID))
	return ""
}

func (f *AgentStateFSM) Handle(fromActor, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	op := stringVal(p, "op", msgType)
	switch op {
	case "transition":
		return f.transition(p)
	case "get_state":
		return marshal(map[string]any{
			"status":           "ok",
			"state":            f.FSMState,
			"transition_count": f.TransitionCount,
		})
	default:
		return marshal(map[string]any{"error": "unknown_op", "op": op})
	}
}

func (f *AgentStateFSM) transition(p map[string]any) string {
	to := stringVal(p, "to", "")
	if to == "" {
		return marshal(map[string]any{"error": "to state is required"})
	}

	if to == "error" || (f.FSMState == "error" && to == "idle") {
		from := f.FSMState
		f.FSMState = to
		f.TransitionCount++
		return marshal(map[string]any{"status": "ok", "from": from, "to": to, "state": f.FSMState})
	}

	allowed := validTransitions()
	validTargets, ok := allowed[f.FSMState]
	if !ok {
		return marshal(map[string]any{"error": fmt.Sprintf("unknown source state: %s", f.FSMState), "state": f.FSMState})
	}
	for _, valid := range validTargets {
		if valid == to {
			from := f.FSMState
			f.FSMState = to
			f.TransitionCount++
			f.IncrCounter(host, "fsm_transitions")
			host.Debug(fmt.Sprintf("AgentStateFSM: %s → %s", from, to))
			return marshal(map[string]any{"status": "ok", "from": from, "to": to, "state": f.FSMState})
		}
	}

	host.Warn(fmt.Sprintf("AgentStateFSM: invalid transition %s → %s", f.FSMState, to))
	return marshal(map[string]any{
		"error": fmt.Sprintf("invalid transition %s → %s", f.FSMState, to),
		"state": f.FSMState,
	})
}

func validTransitions() map[string][]string {
	return map[string][]string{
		"idle":           {"processing"},
		"processing":     {"tool_executing", "responding"},
		"tool_executing": {"processing"},
		"responding":     {"idle"},
		"error":          {"idle"},
	}
}

File: examples/python/apps/miniclaw/memory.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""MemoryActor — scoped KV + TupleSpace memory store.

AuditEventActor — GenEvent fire-and-forget audit trail.
AgentStateFSM — agent lifecycle state machine.
"""

from plexspaces import actor, event_actor, fsm_actor, state, handler, init_handler, host
from .helpers import fire_audit

_VALID_FSM_TRANSITIONS = {
    "idle": {"processing", "tool_executing"},
    "processing": {"tool_executing", "responding", "idle"},
    "tool_executing": {"processing", "idle"},
    "responding": {"idle"},
}


# ---------------------------------------------------------------------------
# MemoryActor
# ---------------------------------------------------------------------------


@actor
class MemoryActor:
    """Scoped memory backed by KV (persistent) and TupleSpace (queryable).

    Supports three scopes: global, agent-scoped, and session-scoped.
    """

    memory_count: int = state(default=0)
    actor_id: str = state(default="")

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        host.process_groups.join("svc:memory")
        host.info(f"MemoryActor init actor_id={self.actor_id}")

    @handler("store_memory")
    def store_memory(self, key: str = "", value: str = "", scope: str = "global", agent_id: str = "", session_id: str = "") -> dict:
        if not key:
            return {"error": "key is required"}
        scoped_key = _scoped_key(scope, agent_id, session_id, key)
        host.kv.put(scoped_key, str(value))
        host.ts.write(["memory", scope, key, str(value)])
        self.memory_count += 1
        fire_audit("memory_stored", f"scope={scope} key={key}")
        return {"status": "ok", "key": key, "scope": scope}

    @handler("recall_memory")
    def recall_memory(self, key: str = "", scope: str = "global", agent_id: str = "", session_id: str = "") -> dict:
        if not key:
            return {"error": "key is required"}
        scoped_key = _scoped_key(scope, agent_id, session_id, key)
        value = host.kv.get(scoped_key)
        return {"status": "ok", "key": key, "value": value, "found": bool(value)}

    @handler("list_memories")
    def list_memories(self, scope: str = "global") -> dict:
        # TupleSpace query: find all ["memory", scope, ?, ?] tuples
        try:
            tuples = host.ts.read_all(["memory", scope, None, None])
            memories = [{"key": t[2], "value": t[3]} for t in tuples if len(t) >= 4]
        except Exception:
            memories = []
        return {"status": "ok", "memories": memories, "scope": scope}

    @handler("delete_memory")
    def delete_memory(self, key: str = "", scope: str = "global", agent_id: str = "", session_id: str = "") -> dict:
        if not key:
            return {"error": "key is required"}
        scoped_key = _scoped_key(scope, agent_id, session_id, key)
        host.kv.delete(scoped_key)
        return {"status": "ok", "key": key}

    @handler("get_stats")
    def get_stats(self) -> dict:
        return {"status": "ok", "memory_count": self.memory_count}


def _scoped_key(scope: str, agent_id: str, session_id: str, key: str) -> str:
    if scope == "agent" and agent_id:
        return f"mem:agent:{agent_id}:{key}"
    if scope == "session" and session_id:
        return f"mem:session:{session_id}:{key}"
    return f"mem:global:{key}"


# ---------------------------------------------------------------------------
# AuditEventActor
# ---------------------------------------------------------------------------


@event_actor
class AuditEventActor:
    """GenEvent actor: receives fire-and-forget audit events and stores them in TupleSpace."""

    event_count: int = state(default=0)
    actor_id: str = state(default="")

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        host.process_groups.join("svc:audit")
        host.info(f"AuditEventActor init actor_id={self.actor_id}")

    @handler("log_event", "cast")
    def log_event(self, event_type: str = "", detail: str = "", timestamp: int = 0) -> None:
        ts = timestamp or host.now_ms()
        try:
            host.ts.write(["audit", event_type, ts, detail])
        except Exception as e:
            host.warn(f"AuditEvent: ts.write failed: {e}")
        self.event_count += 1
        host.debug(f"audit event_type={event_type} detail={detail}")

    @handler("get_stats")
    def get_stats(self) -> dict:
        return {"status": "ok", "event_count": self.event_count}


# ---------------------------------------------------------------------------
# AgentStateFSM
# ---------------------------------------------------------------------------


@fsm_actor(states=["idle", "processing", "tool_executing", "responding"], initial="idle")
class AgentStateFSM:
    """Agent lifecycle FSM: idle → processing → tool_executing → responding → idle."""

    fsm_state: str = state(default="idle")
    transition_count: int = state(default=0)
    actor_id: str = state(default="")

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        host.process_groups.join("svc:agent_fsm")
        host.info(f"AgentStateFSM init actor_id={self.actor_id}")

    @handler("transition")
    def transition(self, to: str = "") -> dict:
        allowed = _VALID_FSM_TRANSITIONS.get(self.fsm_state, set())
        if to not in allowed:
            host.debug(f"FSM: invalid transition {self.fsm_state} → {to}")
            return {"status": "ignored", "from": self.fsm_state, "to": to}
        prev = self.fsm_state
        self.fsm_state = to
        self.transition_count += 1
        host.debug(f"FSM: {prev} → {to}")
        return {"status": "ok", "from": prev, "to": to}

    @handler("get_state")
    def get_state(self) -> dict:
        return {"status": "ok", "state": self.fsm_state, "transitions": self.transition_count}
